### DISCUSSION TITLE
Example of SIMD-like behavior using intrinsics?

### DISCUSSION THREAD

**Discussion by qpwo** [vlang/v#14841] [General]
It is discussed [here](https://github.com/vlang/v/issues/1079#issuecomment-549320821) and [here](https://github.com/vlang/v/issues/1868#issuecomment-528559933), but I'm no c++ wiz and I can't find an example in vlang, so:

Can someone provide a minimal example of broadcasting an operation over an array in vlang?

How could I achieve something like `[1, 2, 3, 4] .* 2` (in whatever syntax)?

**Comment by JalonSolov** [comment]
If you mean "multiply all the elements by 2", that would be `[1, 2, 3, 4].map(it * 2)`.

Otherwise, I don't know what you meant by `.* 2`

**Reply by qpwo** [reply]
Thanks but I'm referring specifically to the optimization where you do the entire operation in one instruction if it's available on the CPU https://en.wikipedia.org/wiki/Single_instruction,_multiple_data

**Reply by qpwo** [reply]
This is like half the reason the language Julia was created as I understand. It's not a common need in server code, but is very useful if you're doing any kind of custom image processing or other matrix transformations/processing

**Reply by JalonSolov** [reply]
Ah.  Then no, V doesn't currently support that.

**Reply by qpwo** [reply]
@dumblob seemed to suggest that it's technically [possible](https://github.com/vlang/v/issues/1868#issuecomment-528559933)? Perhaps I misread that and they were saying it would have to be added to the compiler

**Reply by JalonSolov** [reply]
Like I said... V doesn't _currently_ support it.  It is definitely something that would have to be added to the compiler.

Nobody has done any work on it so far, and I haven't seen it mentioned in any priority lists so far.  Doesn't mean it won't be handled, just no estimates on when work might start.

**Reply by qpwo** [reply]
Cool, fyi SIMD [seems](https://stackoverflow.blog/2020/07/08/improving-performance-with-simd-intrinsics-in-three-use-cases/#:~:text=Best%20versions%20are%205%2D12%20times%20faster%20than%20scalar%20code.) to be 10x faster in ideal cases such as dot product.

I'm not sure (1) whether you'd want it be done explicitly in syntax or have the v compiler do it, or (2) if gcc or clang will automatically do this already somehow or (3) how valued matrix math use cases (video processing, ML, etc) are by the v community. 

But if vector operations / SIMD were supported then it would make V a very compelling language for such cases.

**Reply by qpwo** [reply]
Unfortunately I would have no idea where to begin to help out

**Reply by JalonSolov** [reply]
While you're waiting for SIMD support, you may want to check these links.  :-)

https://github.com/vlang/vsl
https://github.com/vlang/vtl

**Reply by dumblob** [reply]
@qpwo yes it is technically possible (and not as difficult to implement compared to other language stuff which has understandably (much) higher priority).

OTOH you can achieve partial SIMD parallelization already now just by using the `vsl` and `vtl` libraries @JalonSolov suggested (as these libraries are partially written in a way which allows the compiler to get a bit wild about aggressive optimizations incl. SIMD).

There is also `blas` V module/library and other wrappers to known highly optimized math/scientific/vector/matrix/3D/... libraries etc. So I think you have several things to try out and chances are they will serve your purpose as well (or close to that) as if you used direct SIMD intrinsics in V :wink: (from my experience such low-level use of SIMD is very unintuitive and without enough experience there is a high chance there will be either no gains or even slight slow down compared to traditional V code without intrinsics!).

**Comment by Wajinn** [comment]
> While you're waiting for SIMD support, you may want to check these links. --JalonSolov


https://github.com/vlang/vsl
https://github.com/vlang/vtl

> There is also blas V module/library and other wrappers to know... --dumblob

Such good suggestions, maybe needs to make sure it's not buried in the replies and possibly overlooked.
